src/lib/operators/window_function_evaluator.hpp

```diff
@@ -0,0 +1,95 @@
+                          std::vector<ColumnID> init_partition_by_column_ids,
+                          std::vector<ColumnID> init_order_by_column_ids, ColumnID init_function_argument_column_id,
+                          std::shared_ptr<WindowFunctionExpression> init_window_funtion_expression);
+
+  const std::string& name() const override;
+  bool is_output_nullable() const;
+
+  enum class ComputationStrategy {
+    OnePass,
+    SegmentTree,
+  };
+
+  template <typename InputColumnType, WindowFunction window_function>
+  ComputationStrategy choose_computation_strategy() const;
+
+  enum class OperatorSteps : uint8_t {
+    MaterializeIntoBuckets,
+    PartitionAndOrder,
+    Compute,
+    Annotate,
+  };
+
+  struct PerformanceData : public OperatorPerformanceData<OperatorSteps> {
+    void output_to_stream(std::ostream& stream, DescriptionMode description_mode) const override;
+    ComputationStrategy computation_strategy;
+  };
+
+ protected:
+  std::shared_ptr<const Table> _on_execute() override;
+
+  template <typename InputColumnType, WindowFunction window_function>
+  std::shared_ptr<const Table> _templated_on_execute();
+
+  void _on_set_parameters(const std::unordered_map<ParameterID, AllTypeVariant>& parameters) override;
+  std::shared_ptr<AbstractOperator> _on_deep_copy(
+      const std::shared_ptr<AbstractOperator>& copied_left_input,
+      const std::shared_ptr<AbstractOperator>& copied_right_input,
+      std::unordered_map<const AbstractOperator*, std::shared_ptr<AbstractOperator>>& copied_ops) const override;
+
+ private:
+  [[nodiscard]] HashPartitionedData materialize_into_buckets() const;
+  void partition_and_order(HashPartitionedData& buckets) const;
+
+  template <typename InputColumnType, WindowFunction window_function>
+    requires SupportsOnePass<InputColumnType, window_function>
+  void compute_window_function_one_pass(const HashPartitionedData& partitioned_data, auto&& emit_computed_value) const;
+
+  template <typename InputColumnType, WindowFunction window_function>
+    requires SupportsSegmentTree<InputColumnType, window_function>
+  void compute_window_function_segment_tree(const HashPartitionedData& partitioned_data,
+                                            auto&& emit_computed_value) const;
+
+  template <typename OutputColumnType>
+  [[nodiscard]] std::shared_ptr<const Table> annotate_input_table(
+      std::vector<std::pair<pmr_vector<OutputColumnType>, pmr_vector<bool>>> segment_data_for_output_column) const;
+
+  const FrameDescription& frame_description() const;
+
+  std::vector<ColumnID> _partition_by_column_ids;
+  std::vector<ColumnID> _order_by_column_ids;
+  ColumnID _function_argument_column_id;
+  std::shared_ptr<WindowFunctionExpression> _window_function_expression;
+};
+
+}  // namespace window_function_evaluator
+
+using WindowFunctionEvaluator = window_function_evaluator::WindowFunctionEvaluator;
```

Comment:
I think we did it like that so that we would not have to write `window_function_evaluator::` everywhere in the class definition. I think I would not want to have _all_ our stuff in the hyrise namespace directly, so this seemed like the most sane solution. What do you think?